File: IndicatorsTests/Test.Sma.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Skender.Stock.Indicators;
using System;
using System.Collections.Generic;
using System.Linq;

namespace StockIndicators.Tests
{
    [TestClass]
    public class SmaTests : TestBase
    {

        [TestMethod()]
        public void GetSmaTest()
        {
            int lookbackPeriod = 20;
            IEnumerable<SmaResult> results = Indicator.GetSma(history, lookbackPeriod);

            // assertions

            // proper quantities
            // should always be the same number of results as there is history
            Assert.AreEqual(502, results.Count());
            Assert.AreEqual(502 - lookbackPeriod + 1, results.Where(x => x.Sma != null).Count());

            // sample value
            SmaResult r = results.Where(x => x.Index == 502).FirstOrDefault();
            Assert.AreEqual((decimal)251.86, r.Sma);
            Assert.AreEqual((decimal)9.45, r.Mad);
            Assert.AreEqual((double)119.2510, Math.Round((double)r.Mse, 4));
            Assert.AreEqual((double)0.037637, Math.Round((double)r.Mape, 6));
        }


        /* EXCEPTIONS */

        [TestMethod()]
        [ExpectedException(typeof(BadParameterException), "Bad lookback.")]
        public void BadLookback()
        {
            Indicator.GetSma(history, 0);
        }

        [TestMethod()]
        [ExpectedException(typeof(BadHistoryException), "Insufficient history.")]
        public void InsufficientHistory()
        {
            Indicator.GetSma(history.Where(x => x.Index < 10), 10);
        }
    }
}

File: wraps/python/tests/test_sma.py
from .TestBase import TestBase
from SkenderStockIndicators import indicators

class TestSma(TestBase):

    def test_standard(self):
        results = indicators.get_sma(self.history, 20)

        # proper quantities
        # should always be the same number of results as there is history
        self.assertEqual(502, len(results))
        self.assertEqual(483, len(list(filter(lambda x: x.Sma is not None, results))))

        # sample values
        self.assertIsNone(results[18].Sma)
        self.assertEqual(214.5250, round(float(results[19].Sma), 4))
        self.assertEqual(215.0310, round(float(results[24].Sma), 4))
        self.assertEqual(234.9350, round(float(results[149].Sma), 4))
        self.assertEqual(255.5500, round(float(results[249].Sma), 4))
        self.assertEqual(251.8600, round(float(results[501].Sma), 4))

    def test_extended(self):
        results = indicators.get_sma_extended(self.history, 20)

        # proper quantities
        # should always be the same number of results as there is history
        self.assertEqual(502, len(results))
        self.assertEqual(483, len(list(filter(lambda x: x.Sma is not None, results))))

        # sample values
        r = results[501]
        self.assertEqual(251.86, float(r.Sma));
        self.assertEqual(9.45, float(r.Mad));
        self.assertEqual(119.2510, round(float(r.Mse), 4))
        self.assertEqual(0.037637, round(float(r.Mape), 6))

    def test_bad_data(self):
        results = indicators.get_sma_extended(self.history, 15)

        self.assertEqual(502, len(results))

    def test_exceptions(self):
        from System import ArgumentOutOfRangeException
        self.assertRaises(ArgumentOutOfRangeException, indicators.get_sma, self.history, 0)

        from Skender.Stock.Indicators import BadHistoryException
        self.assertRaises(BadHistoryException, indicators.get_sma, self.data_reader.get(9), 10)
